cmd/podman/images/scp.go

```diff
@@ -0,0 +1,410 @@
+// an scp command and then calls podman image load on the remote machine given the new scp filepath
+func execCommand(iden string, uri *urlP.URL, imageFileIn string, imageFileOut string, image string, LoadSave bool) (string, error) {
+	var signers []ssh.Signer
+	passwd, passwdSet := uri.User.Password()
+	if iden != "" {
+		value := iden
+		s, err := t.PublicKey(value, []byte(passwd))
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to read identity %q", value)
+		}
+		signers = append(signers, s)
+		logrus.Debugf("SSH Ident Key %q %s %s", value, ssh.FingerprintSHA256(s.PublicKey()), s.PublicKey().Type())
+	}
+	if sock, found := os.LookupEnv("SSH_AUTH_SOCK"); found { // validate ssh information
+		logrus.Debugf("Found SSH_AUTH_SOCK %q, ssh-agent signer enabled", sock)
+
+		c, err := net.Dial("unix", sock)
+		if err != nil {
+			return "", err
+		}
+		agentSigners, err := agent.NewClient(c).Signers()
+		if err != nil {
+			return "", err
+		}
+
+		signers = append(signers, agentSigners...)
+
+		if logrus.IsLevelEnabled(logrus.DebugLevel) {
+			for _, s := range agentSigners {
+				logrus.Debugf("SSH Agent Key %s %s", ssh.FingerprintSHA256(s.PublicKey()), s.PublicKey().Type())
+			}
+		}
+	}
+	var authMethods []ssh.AuthMethod
+	if len(signers) > 0 {
+		var dedup = make(map[string]ssh.Signer)
+		for _, s := range signers {
+			fp := ssh.FingerprintSHA256(s.PublicKey())
+			if _, found := dedup[fp]; found {
+				logrus.Debugf("Dedup SSH Key %s %s", ssh.FingerprintSHA256(s.PublicKey()), s.PublicKey().Type()) // public key info and validation
+			}
+			dedup[fp] = s
+		}
+		var uniq []ssh.Signer
+		for _, s := range dedup {
+			uniq = append(uniq, s)
+		}
+		authMethods = append(authMethods, ssh.PublicKeysCallback(func() ([]ssh.Signer, error) {
+			return uniq, nil
+		}))
+	}
+	if passwdSet { // proper password validation for ssh connection if necessary
+		authMethods = append(authMethods, ssh.Password(passwd))
+	}
+	if len(authMethods) == 0 {
+		authMethods = append(authMethods, ssh.PasswordCallback(func() (string, error) {
+			pass, err := t.ReadPassword(fmt.Sprintf("%s's login password:", uri.User.Username()))
+			return string(pass), err
+		}))
+	}
+	cfg := &ssh.ClientConfig{ // configure ssh client
+		User:            uri.User.Username(),
+		Auth:            authMethods,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	dial, err := ssh.Dial("tcp", uri.Host, cfg) // dial the client
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to connect")
+	}
+	defer dial.Close()
+
+	session, err := dial.NewSession() // connect to the dialed client
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create new ssh session on %q", uri.Host)
+	}
+	defer session.Close()
+	if err != nil {
+		return "", err
+	}
+
+	if LoadSave { // if we are loading to the remote/ssh client
+		f, _ := os.Open(imageFileIn) // open image file to copy
+		defer f.Close()
+		n, err := scpD.CopyTo(dial, imageFileIn, imageFileOut)
```

Comment:
for remote load, I just added a `...;rm filename` but for save I cannot since session.Run can only do one command per session and I need to copy the file to the local machine before I can delete it. I can investigate adding in a v2?

Reply:
I think we shouldn’t release this leaving gigabytes of unexpected data around — but I’ll defer to Podman maintainers.